### ISSUE TITLE
SCDF Dashboard - Provide search/filter options in jobs, tasks, streams and apps view page

### ISSUE THREAD

**Issue by majorisit** [spring-cloud/spring-cloud-dataflow#3866]
The current version of **scdf dashboard** is not flexible enough to **search/filter** deployed/running components using its name, id, status, or time. Please make enhancement with the current scdf dashboard to provide search capability at the column header level.  Currently, Its not flexible enough to search/filter scdf deployed components due to pagination.

![scdf_dashboard](https://user-images.githubusercontent.com/31349495/78922793-25a3d180-7a65-11ea-8c5a-c8a1d7785707.jpg)

**Comment by siddhantsorann** [comment]
@sabbyanandan Hey, I'd like to give this a shot, as we would love to have this feature for our use case as well. Can you let me know how to go about it?

**Comment by sabbyanandan** [comment]
@siddhantsorann: Excellent! @oodamien can help guide by showing you around the codebase, which hopefully can get you going.

**Comment by sabbyanandan** [comment]
The backend is probably already available. If so, we may have to move this card to the UI repo. @ilayaperumalg/@jvalkeal: could one of you confirm?

**Comment by siddhantsorann** [comment]
Hey, @sabbyanandan any updates on this? You can also let me know if there is some other ticket I can pick up which might be a good starting point.

**Comment by sabbyanandan** [comment]
@ilayaperumalg @jvalkeal @oodamien: Could one of you help guide us here?

**Comment by sabbyanandan** [comment]
I just also realize https://github.com/spring-cloud/spring-cloud-dataflow/issues/3564 could also benefit from this capability.

**Comment by siddhantsorann** [comment]
Yeah, that also seems to have the same context. We, ourselves would love to have this filter as well. Hence eager to pick this up.

**Comment by ilayaperumalg** [comment]
@siddhantsorann Thanks for the interest in this and sorry about the delay in getting back to you on this.

Here are the changes you would need to consider at the backend before you start designing at the UI (especially those queries related to searching batch job executions `from` and `to` etc.,) :

- Introduce the batch job retrieval methods that are necessary on these filtering cases in this [JobService] type. (https://github.com/spring-cloud/spring-cloud-dataflow/blob/497f6371111a59edd9e6fcba2b46eeb7f3d6ab65/spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/batch/JobService.java#L49)
- SCDF uses the [SimpleJobService](https://github.com/spring-cloud/spring-cloud-dataflow/blob/497f6371111a59edd9e6fcba2b46eeb7f3d6ab65/spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/batch/SimpleJobService.java#L64) implementation of the `JobService`. Hence, you need to implement those new methods in here.
- You also need to add new contracts and implement the necessary DAO methods in the DAOs in [here](https://github.com/spring-cloud/spring-cloud-dataflow/tree/master/spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/batch). For instance, the [JdbcSearchableJobExecutionDao](https://github.com/spring-cloud/spring-cloud-dataflow/blob/master/spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/batch/JdbcSearchableJobExecutionDao.java#L49) has some of the similar implementations.

**Comment by siddhantsorann** [comment]
Hey, @ilayaperumalg Thanks for the information. I'll get started upon understanding the code base and changes required.

**Comment by siddhantsorann** [comment]
Hey, @ilayaperumalg I'm planning on starting with a basic search filter on the job executions page. For that, I'll have to implement it in JobService right? And add a new DAO method?

**Comment by ilayaperumalg** [comment]
Yes, You need to start adding the specific filtering query at the DAO and the corresponding retrieval methods in `SimpleJobService`

**Comment by siddhantsorann** [comment]
Hey @ilayaperumalg, with regards to the API for fetch job executions defined [here](https://github.com/spring-cloud/spring-cloud-dataflow/blob/master/spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/controller/JobExecutionThinController.java), should I continue using the same API and add an optional parameter "q" which if not empty, will call the DAO which searches with jobName Like.

**Comment by ilayaperumalg** [comment]
If adding a parameter provides reusability, you can use it.

I suggest you implement a first cut and submit a draft PR. I think it would be good to discuss from there.

**Comment by siddhantsorann** [comment]
Sure. That makes sense. Will do.

**Comment by siddhantsorann** [comment]
Hey, @ilayaperumalg any updates on this? In case someone is unavailable could I please get an ETA on this. I could plan to pick up some tickets accordingly. Thanks.

**Comment by actepotentiel** [comment]
Hello gentlemen, first of all I would like to congratulate you for your excellent work. Filtering job executions is the last feature my team is waiting for to migrate to SCDF. Would it also be possible to include a filter on the job parameters? If this is not a priority for you I would be happy to contribute to the project by developing this feature. Regards.

**Comment by siddhantsorann** [comment]
Hey @actepotentiel ! This was a similar case with my team, we are currently using SCDF and wanted the filters for jobs, which is why I decided to pick this ticket up and contribute. Currently, as part of the PR I have only added a filter to search with job name. Was hoping this would get me a deeper insight into the codebase to pick up various other filters. Although, depending on the priority defined by SCDF maintainers we could divide this further into multiple filters and issues.

**Comment by siddhantsorann** [comment]
Hey @cppwfs @ilayaperumalg For this ticket should I raise a PR for other filters? Like job/task execution ID, date range, status etc?

**Comment by ilayaperumalg** [comment]
Hi @siddhantsorann , 

Sure, please go ahead and submit PRs. Thanks for your interest in contributing!

**Comment by siddhantsorann** [comment]
@cppwfs @ilayaperumalg Please have a look at the above PR when possible.

**Comment by siddhantsorann** [comment]
@sabbyanandan Does it also make sense to include a filter for dslText and/or definition for tasks as part of this issue? Currently, filter is available only by name.

**Comment by sabbyanandan** [comment]
@siddhantsorann: Sure, why not. It is a String literal similar to the Task definition name anyway. Let's add it.

**Comment by siddhantsorann** [comment]
@cppwfs @ilayaperumalg If we add more filters to job executions, do we need to combine them as well? For example date filter + name filter. Currently, you can only filter by either. Also, if we do that should we consider using queryDSL and predicates? As the repo call becomes dynamic based on all parameters provided by the user in the API request. 
Otherwise, we might have too many if-else based on all parameters to figure out which repo call to use.

**Comment by cppwfs** [comment]
Good idea.  But for now let us get an MVP where we handle each  filter independently.

**Comment by siddhantsorann** [comment]
So for the next filter, I'll be working on filtering out job execution Id by task-execution id. Should this be a separate API call in which only task execution id is sent as a parameter?

**Comment by siddhantsorann** [comment]
@cppwfs @ilayaperumalg Are there any other filters that we need to create? For tasks or batch jobs?

**Comment by cppwfs** [comment]
@siddhantsorann Absolutely :-)
Here they are:
https://github.com/spring-cloud/spring-cloud-dataflow/issues/4165
https://github.com/spring-cloud/spring-cloud-dataflow/issues/4166

**Comment by cppwfs** [comment]
Re Opening because these solutions are mutually exclusive to each other.   You can search for exec_id or instance_id or start / end dates.   But they can't be combined.
